mlir-sycl/lib/Conversion/SYCLToGPU/SYCLToGPU.cpp

```diff
@@ -0,0 +1,223 @@
+      static_cast<Value>(rewriter.create<arith::ConstantIndexOp>(loc, 0));
+  const auto res = static_cast<Value>(
+      rewriter.replaceOpWithNewOp<memref::LoadOp>(op, alloca, zero));
+  const auto argumentTypes =
+      rewriter.getTypeArrayAttr({MemRefType::get(1, resTy, {}, 4), getIndexTy});
+  const auto functionName = rewriter.getAttr<FlatSymbolRefAttr>("operator[]");
+  // Initialize
+  for (int64_t i = 0; i < dimensions; ++i) {
+    const auto index = static_cast<Value>(
+        rewriter.create<arith::ConstantIntOp>(loc, i, getIndexTy));
+    const auto val = static_cast<Value>(rewriter.create<arith::IndexCastOp>(
+        loc, targetIndexTy,
+        getDimension(rewriter, loc, opName, dimensionAttrname, i)));
+    const auto ptr = createGetOp(rewriter, loc, dimMtTy, res, index,
+                                 argumentTypes, functionName);
+    rewriter.create<memref::StoreOp>(loc, val, ptr, zero);
+  }
+}
+
+/// Converts n-dimensional operations to operations of name \p opName, from the
+/// GPU dialect.
+///
+/// There are three possible cases here:
+/// 1. No argument is passed (see convertToFullObject())
+/// 2. A constant argument is passed (see converWithConstIndex())
+/// 3. A non-constant argument is passed (see convertWithVarIndex())
+void rewrite(StringRef opName, StringRef dimensionAttrName, Operation *op,
+             ValueRange operands, ConversionPatternRewriter &rewriter) {
+  switch (op->getNumOperands()) {
+  case 0: {
+    const auto dimensions = getDimensions(op->getResultTypes()[0]);
+    convertToFullObject(rewriter, opName, dimensionAttrName, op, dimensions);
+    break;
+  }
+  case 1: {
+    const auto dimension = operands[0];
+    if (const auto definingOp = dimension.getDefiningOp<arith::ConstantOp>()) {
+      convertWithConstIndex(rewriter, opName, dimensionAttrName, op,
+                            definingOp);
+    } else {
+      // TODO: Do not rely in a default number of dimensions; take into
+      // account kernel dimensionality. This should be available in the parent
+      // function.
+      constexpr int64_t defaultDimensions{3};
+      convertWithVarIndex(rewriter, opName, dimensionAttrName, op,
+                          defaultDimensions, dimension);
+    }
+    break;
+  }
+  default:
+    llvm_unreachable("Invalid cardinality");
+  }
+}
+
+/// Converts n-dimensional operations of type \tparam OpTy to operations of type
+/// \tparam GPUOpTy.
+///
+/// Work is offloaded to the rewrite function above.
+template <typename OpTy, typename GPUOpTy>
+class GridOpPattern final : public OpConversionPattern<OpTy> {
+public:
+  using OpConversionPattern<OpTy>::OpConversionPattern;
+
+  LogicalResult
+  matchAndRewrite(OpTy op, typename OpTy::Adaptor opAdaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    constexpr auto opName = GPUOpTy::getOperationName();
+    ::rewrite(opName,
+              GPUOpTy::getDimensionAttrName({opName, rewriter.getContext()}),
```

Comment:
What's the benefit of passing `dimensionAttrName` around, instead of querying when needed?

Reply:
Although it is in the end the same name for each operation, there is no guarantee on this: there is no `GPUDialect::getDimensionAttrName`, but the `getDimensionAttrName` is defined for each operation. As we only know the type of the operation in the templated function, we need to pass it around like that unfortunately.